mediapipe/graphs/instantmotiontracking/calculators/matrices_manager.cc

```diff
@@ -0,0 +1,221 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <memory>
+#include <cmath>
+#include "Eigen/Dense"
+#include "Eigen/src/Core/util/Constants.h"
+#include "Eigen/src/Geometry/Quaternion.h"
+#include "absl/memory/memory.h"
+#include "absl/strings/str_cat.h"
+#include "absl/strings/str_join.h"
+#include "mediapipe/framework/calculator_framework.h"
+#include "mediapipe/framework/port/ret_check.h"
+#include "mediapipe/framework/port/status.h"
+#include "mediapipe/graphs/object_detection_3d/calculators/box.h"
+#include "mediapipe/graphs/object_detection_3d/calculators/model_matrix.pb.h"
+#include "mediapipe/graphs/instantmotiontracking/calculators/transformations.h"
+
+namespace mediapipe {
+
+using Matrix4fRM = Eigen::Matrix<float, 4, 4, Eigen::RowMajor>;
+
+constexpr char kFinalTranslationsTag[] = "TRANSLATION_DATA";
+constexpr char kIMUDataTag[] = "IMU_DATA";
+constexpr char kUserRotationsTag[] = "USER_ROTATIONS";
+constexpr char kUserScalingsTag[] = "USER_SCALINGS";
+constexpr char kModelMatricesTag[] = "MODEL_MATRICES";
+
+
+// Intermediary for rotation and translation data to model matrix usable by
+// gl_animation_overlay_calculator
+//
+// Input:
+//  IMU_DATA - float[3] of [roll, pitch, yaw] of device
+//  USER_ROTATIONS - UserRotations with corresponding radians of rotation
+//  TRANSLATION_DATA - All stickers final translation data (vector of
+//  Translation objects)
+// Output:
+//  MODEL_MATRICES - TimedModelMatrixProtoList of all objects to render
+//
+// Example config:
+// node{
+//  calculator: "MatricesManagerCalculator"
+//  input_stream: "IMU_DATA:imu_data"
+//  input_stream: "USER_ROTATIONS:user_rotation_data"
+//  input_stream: "TRANSLATION_DATA:final_translation_data"
+//  output_stream: "MODEL_MATRICES:model_matrices"
+// }
+
+class MatricesManagerCalculator : public CalculatorBase {
+private:
+  // (68 degrees, 4:3 for Pixel 4)
+  const float vertical_fov_radians_ = (68.0f) * M_PI / 180.0f;
+  const float aspect_ratio_ = (4.0f / 3.0f);
+  // initial Z value (-98 is just in visual range for OpenGL render)
+  const float initial_z_ = -98;
+
+ public:
+  static ::mediapipe::Status GetContract(CalculatorContract* cc);
+  ::mediapipe::Status Open(CalculatorContext* cc) override;
+  ::mediapipe::Status Process(CalculatorContext* cc) override;
+  Matrix4fRM generateEigenModelMatrix(Eigen::Vector3f translation_vector,
+    Eigen::Matrix3f rotation_submatrix);
+  Eigen::Vector3f generateTranslationVector(Anchor tracked_anchor, float user_scaling_increment);
+  Eigen::Matrix3f generateRotationSubmatrix(float roll, float pitch, float yaw, float user_rotation_radians);
+
+  float getUserScaling(std::vector<UserScaling> scalings, int sticker_id) {
```

Comment:
For a small number of stickers, this is probably optimal, but if we end up wanting to support a decent number of stickers, then we might want to change this lookup.  I'm assuming the number of stickers we want to handle at once will stay capped to something relatively low, but if that's not the case, maybe add a quick TODO here to adjust the lookup function later.